src/dune/string_with_vars.ml

```diff
@@ -229,68 +229,108 @@ end
 
 open Private
 
-let partial_expand :
-      'a.    t -> mode:'a Mode.t -> dir:Path.t -> f:Value.t list option expander
-      -> 'a Partial.t =
- fun ({ template; syntax_version } as t) ~mode ~dir ~f ->
-  let commit_text acc_text acc =
-    let s = concat_rev acc_text in
-    if s = "" then
-      acc
-    else
-      Text s :: acc
-  in
-  let rec loop acc_text acc items =
-    match items with
-    | [] -> (
-      match acc with
-      | [] -> Partial.Expanded (Mode.string mode (concat_rev acc_text))
-      | _ ->
-        let template =
-          { template with parts = List.rev (commit_text acc_text acc) }
-        in
-        Unexpanded { template; syntax_version } )
-    | Text s :: items -> loop (s :: acc_text) acc items
-    | (Var var as it) :: items -> (
-      match f var syntax_version with
-      | Some (([] | _ :: _ :: _) as e) when not template.quoted ->
-        invalid_multivalue var e
-      | Some t -> loop (Value.L.concat ~dir t :: acc_text) acc items
-      | None -> loop [] (it :: commit_text acc_text acc) items )
-  in
-  match template.parts with
-  | [] -> Partial.Expanded (Mode.string mode "")
-  | [ Text s ] -> Expanded (Mode.string mode s)
-  | [ Var var ] when not template.quoted -> (
-    match f var syntax_version with
-    | None -> Partial.Unexpanded t
-    | Some e ->
-      Expanded
-        ( match Mode.value mode e with
-        | None -> invalid_multivalue var e
-        | Some s -> s ) )
-  | _ -> loop [] [] template.parts
-
-let expand t ~mode ~dir ~f =
-  match
-    partial_expand t ~mode ~dir ~f:(fun var syntax_version ->
-        match f var syntax_version with
-        | None -> (
-          match var.syntax with
-          | Percent ->
-            if Var.is_macro var then
-              User_error.raise ~loc:var.loc
-                [ Pp.textf "Unknown macro %s" (Var.describe var) ]
-            else
-              User_error.raise ~loc:var.loc
-                [ Pp.textf "Unknown variable %S" (Var.name var) ]
-          | Dollar_brace
-          | Dollar_paren ->
-            Some [ Value.String (string_of_var var) ] )
-        | s -> s)
-  with
-  | Partial.Expanded s -> s
-  | Unexpanded _ -> assert false
+module type S = sig
+  type 'a app
+
+  val expand :
+       t
+    -> mode:'a Mode.t
+    -> dir:Path.t
+    -> f:Value.t list option app expander
+    -> 'a app
+
+  val partial_expand :
+       t
+    -> mode:'a Mode.t
+    -> dir:Path.t
+    -> f:Value.t list option app expander
+    -> 'a Partial.t app
+end
+
+module Make (A : Applicative_intf.S1) = struct
+  module App = Applicative.Make (A)
```

Comment:
For many applicatives `all` can be more efficient than fold of `let+`.
I think it's fine to have `all` as a field of `A` directly instead of deriving it here. (it's not a big deal at this point, though)

Reply:
Agreed. However, I'd prefer working on applying this change to expanding actions within the applicative. This is more important for the overall progress I feel.